GH-913: Support for ErrorMessageSendingRecoverer

garyrussell opened the pull request:
Needs https://github.com/spring-projects/spring-integration/pull/2091 + polishing commits mentioned

Resolves #913

mbogoevici commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +277,55 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+	}
+
+	protected String errorsBaseName(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<?> extendedConsumerProperties) {
```
Does the argument need to be `ExtendedConsumerProperties` or can it be `ConsumerProperties` ? Seems we are just using instanceIndex which is coming from the base class.

mbogoevici commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +277,55 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+		return recoverer;
+	}
+
+	protected String getErrorChannelName(ConsumerDestination destination, String group,
```
I suspect that the main use case would be to forward all error messages to `errorChannel` (I believe this is what we originally discussed) - maybe we should make that the default and allow users to configure it?

garyrussell commented:
@mbogoevici Reworked for some of the discussions on our call last week.

- refactored infrastructure to be mostly in the abstract class
- bridge each error channel to the global error channel
- added a default handler that throws the exception to the container if there are no user (or recovering) handler subsribed - it's configured to be the last handler so its always run after the user handlers
- properly unsubscribe/destroy the error infrastructure beans on unbind
- unfortunately, I couldn't avoid the binder impl having to call `registerErrorInfrastructure` because we need the final group name (after anonymous determination)

artembilan commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +295,125 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+	 * @return the handler.
+	 */
+	protected MessageHandler getDefaultErrorMessageHandler(final SubscribableChannel errorChannel) {
+		class FinalRethrowingErrorMessageHandler implements MessageHandler, Orderable {
```
Maybe would be better to have this as a separate `static` class with the `errorChannel` as ctor arg?

artembilan commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +295,125 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+		String errorChannelName = errorsBaseName(destination, group, extendedConsumerProperties);
+		SubscribableChannel errorChannel;
+		if (getApplicationContext().containsBean(errorChannelName)) {
+			errorChannel = getApplicationContext().getBean(errorChannelName, SubscribableChannel.class);
```
Well, there is no guarantee that end-user always declare error channel as `PublishSubscribeChannel`.
Or is it a part of binding service as well here?

artembilan commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -366,4 +522,62 @@ public boolean isRunning() {
 		}
 	}
+
+	private interface SubscriberAwareChannel {
```
Looks like nobody implements this interface.
I guess the idea was to make that `ErrorChannel` below to follow this contract.

Although looking to that `FinalRethrowingErrorMessageHandler` it really doesn't matter if use has subscribers or not.
You claim there like `run last if the user has subscribers too`.

Nevertheless maybe that would be much better to **remove** that `FinalRethrowingErrorMessageHandler` from the `ErrorChannel` just in the moment use calls its `subscribe()`.

artembilan commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +295,125 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+		beanFactory.registerSingleton(errorMessageHandlerName, handler);
+		beanFactory.initializeBean(handler, errorMessageHandlerName);
+		errorChannel.subscribe(handler);
+		return new ErrorInfrastructure(errorChannel, recoverer, handler);
```
I'm not sure yet why I would need such a complex object in return, but  believe some JavaDocs on the matter will be helpful for any target `Binder` implementators.
Thanks

mbogoevici commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -216,6 +226,33 @@ protected void afterUnbindProducer(ProducerDestination destination, P producerPr
+							(ExtendedConsumerProperties<?>) properties);
+					if (getApplicationContext().containsBean(recoverer)) {
+						((DefaultSingletonBeanRegistry) getApplicationContext().getBeanFactory())
+							.destroySingleton(recoverer);
```
I'm not sure if that will work - given that the binding is not a bean (which has been a discussion point for quite a while).

garyrussell commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -258,6 +295,125 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+	 * @return the implementation - may be null.
+	 */
+	protected ErrorMessageStrategy getErrorMessageStrategy() {
+		return null;
```
That happens at the SI level - see this code in the binder...

		if (errorMessageStrategy == null) {
			recoverer = new ErrorMessageSendingRecoverer(errorChannel);
		}
		else {
			recoverer = new ErrorMessageSendingRecoverer(errorChannel, errorMessageStrategy);
		}

garyrussell commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -366,4 +522,62 @@ public boolean isRunning() {
+
+	protected static class ErrorInfrastructure {
+
+		private final SubscribableChannel errroChannel;
```
Because we have to subscribe a handler to it.

viniciusccarvalho commented on spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java:
```diff
@@ -283,6 +292,163 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
+	 * @return the implementation - may be null.
+	 */
+	protected ErrorMessageStrategy getErrorMessageStrategy() {
+		return null;
```
Not the right place to start this, but I just wanted to raise a concern I have on the way we handle names. It feels a bit odd that we have so many places to handle names for consumer groups, partitions, errors. And sometimes binders may not even use them. I wonder if another approach, less invasive to the base class, such as a strategy pattern could be applied, so any time we add a new feature that require a custom name, we don't need to change the base classes of the project. Happy to discuss this on our next call

viniciusccarvalho commented:
Squashed and merged via 3c2c7da5085559165cd4c1bc4aef44434a39f990
